src/lib/transition_frontier/transition_frontier0.ml

```diff
@@ -0,0 +1,645 @@
+   *  Modifications are in-place
+  *)
+  let move_root t soon_to_be_root_node : node =
+    let root_node = Hashtbl.find_exn t.table t.root in
+    let root_breadcrumb = root_node.breadcrumb in
+    let root = root_breadcrumb |> Breadcrumb.staged_ledger in
+    let soon_to_be_root =
+      soon_to_be_root_node.breadcrumb |> Breadcrumb.staged_ledger
+    in
+    let children =
+      List.map soon_to_be_root_node.successor_hashes ~f:(fun h ->
+          (Hashtbl.find_exn t.table h).breadcrumb |> Breadcrumb.staged_ledger
+          |> Inputs.Staged_ledger.ledger )
+    in
+    let root_ledger = Inputs.Staged_ledger.ledger root in
+    let soon_to_be_root_ledger = Inputs.Staged_ledger.ledger soon_to_be_root in
+    let soon_to_be_root_merkle_root =
+      Ledger.merkle_root soon_to_be_root_ledger
+    in
+    Ledger.commit soon_to_be_root_ledger ;
+    let root_ledger_merkle_root_after_commit =
+      Ledger.merkle_root root_ledger
+    in
+    [%test_result: Ledger_hash.t]
+      ~message:
+        "Merkle root of soon-to-be-root before commit, is same as root \
+         ledger's merkle root afterwards"
+      ~expect:soon_to_be_root_merkle_root root_ledger_merkle_root_after_commit ;
+    let new_root =
+      Breadcrumb.create soon_to_be_root_node.breadcrumb.transition_with_hash
+        (Inputs.Staged_ledger.replace_ledger_exn soon_to_be_root root_ledger)
+    in
+    let new_root_node = {soon_to_be_root_node with breadcrumb= new_root} in
+    let new_root_hash =
+      soon_to_be_root_node.breadcrumb.transition_with_hash.hash
+    in
+    Ledger.remove_and_reparent_exn soon_to_be_root_ledger
+      soon_to_be_root_ledger ~children ;
+    Hashtbl.remove t.table t.root ;
+    Hashtbl.set t.table ~key:new_root_hash ~data:new_root_node ;
+    t.root <- new_root_hash ;
+    new_root_node
+
+  (* Adding a breadcrumb to the transition frontier is broken into the following steps:
+   *   1) attach the breadcrumb to the transition frontier
+   *   2) calculate the distance from the new node to the parent and the
+   *      best tip node
+   *   3) set the new node as the best tip if the new node has a greater length than
+   *      the current best tip
+   *   4) move the root if the path to the new node is longer than the max length
+   *       I   ) find the immediate successor of the old root in the path to the
+   *             longest node (the heir)
+   *       II  ) find all successors of the other immediate successors of the
+   *             old root (bads)
+   *       III ) cleanup bad node masks, but don't garbage collect yet
+   *       IV  ) move_root the breadcrumbs (rewires staged ledgers, cleans up heir)
+   *       V   ) garbage collect the bads
+   *       VI  ) grab the new root staged ledger
+   *       VII ) notify the consensus mechanism of the new root
+   *       VIII) if commit on an heir node that just emitted proof txns then
+   *             write them to snarked ledger
+  *)
+  let add_breadcrumb_exn t breadcrumb =
```

Comment:
for changes to this function, we try to very carefully document the steps since a lot is going on -- can you add comment about returning the extension diff information?